ReactSkia/sdk/FollyTimer.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+* Copyright (C) 1994-2022 OpenTV, Inc. and Nagravision S.A.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+#include <algorithm>
+
+#include "FollyTimer.h"
+
+namespace rns {
+namespace sdk {
+
+Timer::Timer(double duration,
+           bool repeats,
+           std::function<void()> cb,
+           bool autostart)
+           : targetDuration_(duration),
+             repeats_(repeats),
+             cb_(cb),
+             timerThread_("FollyTimerThread") {
+
+  timerThread_.getEventBase()->waitUntilRunning();
+  RNS_LOG_DEBUG("[" << this << "] Created timer with duration:" << targetDuration_ << " repeats:" << repeats_ << " autostart:" << autostart);
+  if(autostart) start();
+
+}
+
+void Timer::start() {
+  RNS_LOG_DEBUG("["<< this << "] Schedule timer for duration:" << targetDuration_);
+
+  if(cb_ == nullptr) {
+    RNS_LOG_ERROR("No callback registered with timer, ignore scheduling of timer");
+    return;
+  }
+
+  if(timerCallback_.isScheduled()) {
+    RNS_LOG_WARN("Timer is already been scheduled for duration:" << targetDuration_);
+    return;
+  }
+
+  auto timeoutCb = std::bind(&Timer::scheduleTimerTimeout, this);
+  // For super fast, on-off timers, just enqueue them immediately rather than waiting
+  if(targetDuration_ <= 1) {
+    timerThread_.getEventBase()->runInEventBaseThread(timeoutCb);
+    return;
+  }
+
+  timerCallback_.cb = timeoutCb;
+  timerThread_.getEventBase()->runInEventBaseThread(
+     [this]() {
+       std::scoped_lock lock(timerCallback_.cbLock);
+       if(timerCallback_.cb) {
+         HHWheelTimer& wheelTimer = timerThread_.getEventBase()->timer();
+         wheelTimer.scheduleTimeout(&timerCallback_, std::chrono::milliseconds(static_cast<unsigned long long>(targetDuration_)));
+       }
+     }
+  );
+  return;
+
+}
+
+void Timer::reschedule(double duration, bool repeats) {
+  RNS_LOG_DEBUG("["<< this << "] Reschedule timer for duration:" << duration << " repeats:" << repeats);
+
+  if(cb_ == nullptr) {
+    RNS_LOG_ERROR("No callback registered with timer, ignore scheduling of timer");
+    return;
+  }
+
+  repeats_ = repeats;
+  if((duration == targetDuration_) && timerCallback_.isScheduled()) {
+    RNS_LOG_WARN("Timer is already scheduled for this duration:" << targetDuration_);
+    return;
+  }
+
+  targetDuration_ = duration;
+
+  if(timerCallback_.cb == nullptr)
+     timerCallback_.cb = std::bind(&Timer::scheduleTimerTimeout, this);
+
```

Comment:
if already scheduled and target duration < 1 , cancel previous timeout and fire immediately.